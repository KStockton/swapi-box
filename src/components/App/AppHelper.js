

function fetchData(urlText){
  let url 

  if(urlText === 'people/' ){
    url = `https:/swapi.co/api/${urlText}`
  } else {
    url = urlText
  }
  return fetch(url)
    .then(response => {
      if(!response.ok){
        throw Error('Error fetching data')
      } else {
        return response.json()
      }
})
}

function getSpecies(characters){
  const completeSpeciesPromise = characters.map(character =>{

    return fetchData(character.species[0])         
           .then(species => ({...character, species}))
  })
  return Promise.all(completeSpeciesPromise)
}

 function getHomeWorld(characterBios){
  const homeWorld = characterBios.map(bio => {
    return fetchData(bio.homeworld)
           .then(homeworlds => ({
             name: bio.name, 
             homeworld: homeworlds.name, 
             population: homeworlds.population, 
             species: bio.species.name, 
             language: bio.species.language }))
  })
  return Promise.all(homeWorld)
}

function getResidents(planets){
  const residents = planets.map(residentOnly => {
 
    if(residentOnly.residents.length > 0) {
      for (let i = 0; i < residentOnly.residents.length; i++)
      return fetchData(residentOnly.residents[i])

    .then(results => (
      {
        name: residentOnly.name, 
        terrain: residentOnly.terrain, 
        population: residentOnly.population, 
        climate: residentOnly.climate, 
        residents: results.name
      }
    )
     ) } else {
          return {
            name: residentOnly.name, 
            terrain: residentOnly.terrain, 
            population: residentOnly.population, 
            climate: residentOnly.climate, 
            residents: `NA`
          }
         }
         return 
       }) 
  return Promise.all(residents)
}


export {
  getHomeWorld,
  getSpecies,
  fetchData,
  getResidents,
}
