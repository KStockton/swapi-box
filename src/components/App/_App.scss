body {
  margin: 0;
  padding: 0;
  background-image: url('https://wallpaperfm.com/img/original/d/e/8/16857.jpg');
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center top;
  font-family: "Roboto", "Orbitron";
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
  pointer-events: none;
}

.App-header {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: stretch;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
