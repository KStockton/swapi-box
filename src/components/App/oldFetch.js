// this.setState({ isLoading: !this.state.isLoading})
// fetchData('https://swapi.co/api/people/')
//  .then(data => data.results.map(person => {
//    const personSpeciesInfo = fetchData(...person.species)
//  .then(species =>  ({ species: species.name, language: species.language}))
//    const personHomeworld = fetchData(person.homeworld)
//  .then(homeworld => ({ homeworld: homeworld.name, population: homeworld.population}))
//    Promise.all([personSpeciesInfo, personHomeworld])
//  .then(personInfo =>{
//      let people = [{...personInfo[0], ...personInfo[1], name: person.name}]
//      this.setState({people})
//      return people
//    })
//  }
//  )
//  )
// }

// function fetchData(url){
//   return fetch(url)
//     .then(response => {
//       if(!response.ok){
//         throw Error('Error fetching data')
//       } else {
//         return response.json()
//       }
// })
// }