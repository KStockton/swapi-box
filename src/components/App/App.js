import React, { Component } from 'react';
import { filmInfo, getSpecies, fetchData, getHomeWorld, onlyPeople } from './AppHelper.js'
import Scroll from '../Scroll/Scroll.js'
import logo from '../../logo.svg';
import './_App.scss';
import Controls from '../Controls/Controls.js'
import CardContainer from '../CardContainer/CardContainer.js'
import { CSSTransition, TransitionGroup } from 'react-transition-group';

export default class App extends Component {
  constructor(){
    super()

  this.state = {
    filmText: {},
    isLoading: true,
    category: '',
    people: [],
    vehicles: '',
    planets: ''
  }
}


componentDidMount = () => {
  const filmNumber = Math.floor(Math.random() * 7) + 1
  const url = `https://swapi.co/api/films/${filmNumber}/`
  fetch(url)
  .then(response => response.json())
  .then(result => filmInfo(result))
  .then(filmText => this.setState(
    { 
      filmText,
      isLoading: false 
    }))
  .catch(error => console.log(error))

}


handleCategory = event => {
  let category = event.target.name
  this.setState({category})
  this.handleFetch(category)
}

handleFetch(usercategory){

  switch(usercategory) {
    case 'people':
      this.getPeople();
      break;
    case 'vehicles':
      this.getVehicles()
      break;
    case 'planets':
      this.getPlanets()
      break;
    default: return null
  }
}

getPeople = () => {
  this.setState({isLoading: !this.state.isLoading})
  fetchData('people/')
  .then(characters => getSpecies(characters.results))
  .then(charactersData => getHomeWorld(charactersData))
  .then(peopleResult => {
      let people = onlyPeople(peopleResult)
      this.setState({ people, isLoading: false})
    })
   
  
}

getPlanets(){

}

getVehicles(){

}



render() {

  const categoryStatus = this.state.category === ''
console.log(this.state.people)
  
  let initialDisplay
  if(this.state.isLoading)
   initialDisplay = 
                  <section>
                    <img src={logo} className="App-logo" alt="logo"/>
                    <h5>Loading have faith...</h5>
                  </section>
 
  return (
    <div className="App">
      <header className="App-header">
        <Controls handleCategory={this.handleCategory}/>
      </header>

      {categoryStatus ? <Scroll filmText={this.state.filmText} />: 
      <CSSTransition
      in={this.state.isLoading}
      appear={true}
      timeout={600}
      classNames="fade"
      >
      
      <CardContainer category={this.state[this.state.category]}/> 
      </CSSTransition>
      }
      <footer>
      {initialDisplay}

      </footer>
    </div>
  );
  }
}
