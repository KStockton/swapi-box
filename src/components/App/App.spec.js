import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

const mockHandleFetch = jest.fn();

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />); 
  });
  it('should match the snatpshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
    
  it('should have a default state', () => {
    const defaultState = {
      filmText: {},
      error: '',
      isLoading: true,
      category: '',
      people: [],
      vehicles: [],
      planets: []
    };
    expect(wrapper.state()).toEqual(defaultState);
  });
    
  it('should update state when handleCategory is called', () => {
    const event = { target: { name: 'people'}};
    wrapper.instance().handleCategory(event);
    expect(wrapper.state('category')).toEqual('people');
  });

  it('should render is loading if the isLoading is true', () => {
    wrapper.setState({isLoading: true});
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Scroll component when the category is an empty string', () => {
    wrapper.setState({category: ""});
    expect(wrapper).toMatchSnapshot();
  });

  xit('should call handleFetch when handleCategory is invoked', () => {
    let mockEvent = { target: { name: 'people'}};
    wrapper.instance().handleCategory(mockEvent);
    expect(mockHandleFetch).toBeCalled();
  });
    
});


