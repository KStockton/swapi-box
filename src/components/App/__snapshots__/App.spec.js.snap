// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snatpshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <section>
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <h5>
        Loading have faith...
      </h5>
    </section>
    <Controls
      handleCategory={[Function]}
    />
    <Favorites />
  </header>
  <Scroll
    filmText={Object {}}
  />
  <footer />
</div>
`;

exports[`App should render Scroll component when the category is an empty string 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <section>
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <h5>
        Loading have faith...
      </h5>
    </section>
    <Controls
      handleCategory={[Function]}
    />
    <Favorites />
  </header>
  <Scroll
    filmText={Object {}}
  />
  <footer />
</div>
`;

exports[`App should render is loading if the isLoading is true 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <section>
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <h5>
        Loading have faith...
      </h5>
    </section>
    <Controls
      handleCategory={[Function]}
    />
    <Favorites />
  </header>
  <Scroll
    filmText={Object {}}
  />
  <footer />
</div>
`;
